Framework/API/test/DetectorInfoTest.h

```diff
@@ -0,0 +1,318 @@
+    TS_ASSERT_EQUALS(detInfo.rotation(0), rot);
+    TS_ASSERT_EQUALS(detInfo.rotation(1), rot);
+    TS_ASSERT_EQUALS(detInfo.rotation(2), rot);
+    TS_ASSERT_EQUALS(detInfo.rotation(3), rot);
+    TS_ASSERT_EQUALS(detInfo.rotation(4), rot);
+    TS_ASSERT_EQUALS(detInfo.sourcePosition(), V3D(0.0, 0.0, 20.0));
+    TS_ASSERT_EQUALS(detInfo.samplePosition(), V3D(0.0, 0.0, 0.0));
+    TS_ASSERT_EQUALS(detInfo.position(0), V3D(0.0, -0.1, -5.0));
+
+    // For additional verification we do *not* use detInfo, but make sure that
+    // the changes actually affected the workspace.
+    const auto &clone = m_workspace.clone();
+    const auto &info = clone->detectorInfo();
+    TS_ASSERT_EQUALS(info.sourcePosition(), V3D(0.0, 0.0, 20.0));
+    TS_ASSERT_EQUALS(info.samplePosition(), V3D(0.0, 0.0, 0.0));
+    TS_ASSERT_EQUALS(info.position(0), V3D(0.0, -0.1, -5.0));
+
+    detInfo.setRotation(*root, oldRot);
+  }
+
+  void test_detectorIDs() {
+    WorkspaceTester workspace;
+    int32_t numberOfHistograms = 5;
+    size_t numberOfBins = 1;
+    workspace.init(numberOfHistograms, numberOfBins, numberOfBins - 1);
+    for (int32_t i = 0; i < numberOfHistograms; ++i)
+      workspace.getSpectrum(i)
+          .setSpectrumNo(static_cast<int32_t>(numberOfHistograms) - i);
+    bool includeMonitors = false;
+    bool startYNegative = true;
+    const std::string instrumentName("SimpleFakeInstrument");
+    InstrumentCreationHelper::addFullInstrumentToWorkspace(
+        workspace, includeMonitors, startYNegative, instrumentName);
+    // Check that *workspace* does not have sorted IDs.
+    TS_ASSERT_EQUALS(workspace.getDetector(0)->getID(), 5);
+    TS_ASSERT_EQUALS(workspace.getDetector(1)->getID(), 4);
+    TS_ASSERT_EQUALS(workspace.getDetector(2)->getID(), 3);
+    TS_ASSERT_EQUALS(workspace.getDetector(3)->getID(), 2);
+    TS_ASSERT_EQUALS(workspace.getDetector(4)->getID(), 1);
+    const auto &info = workspace.detectorInfo();
+    const auto &ids = info.detectorIDs();
+    auto sorted_ids(ids);
+    std::sort(sorted_ids.begin(), sorted_ids.end());
+    // The ids we get from DetectorInfo should be sorted.
+    TS_ASSERT_EQUALS(ids, sorted_ids);
+  }
+
+private:
+  WorkspaceTester m_workspace;
+  WorkspaceTester m_workspaceNoInstrument;
+
+  std::unique_ptr<MatrixWorkspace> makeWorkspace(size_t numSpectra) {
+    auto ws = Kernel::make_unique<WorkspaceTester>();
+    ws->initialize(numSpectra, 1, 1);
+    auto inst = boost::make_shared<Instrument>("TestInstrument");
+    ws->setInstrument(inst);
+    auto &pmap = ws->instrumentParameters();
+    for (size_t i = 0; i < ws->getNumberHistograms(); ++i) {
+      auto det = new Detector("pixel", static_cast<detid_t>(i), inst.get());
+      inst->add(det);
+      inst->markAsDetector(det);
+      ws->getSpectrum(i).addDetectorID(static_cast<detid_t>(i));
+      if (i % 2 == 0)
+        pmap.addBool(det->getComponentID(), "masked", true);
+    }
+    return std::move(ws);
+  }
+};
+
```

Comment:
Would this be a good time to add a performance test suite? At least to help spot regressions.